source/pervasive/string.rs

```diff
@@ -25,47 +25,83 @@ pub const fn new_strlit<'a>(s: &'a str) -> StrSlice<'a> {
+    
+    /// The len() function in rust returns the byte length. 
+    /// It is more useful to talk about the length of characters and therefore this function was added. 
+    /// Please note that this function counts the unicode variation selectors as characters. 
+    /// Warning: O(n)
     #[verifier(external_body)]
-    pub fn len(&self) -> (l: usize)
-        ensures self.is_ascii() ==> l as nat === self.bytes().len()
+    pub fn unicode_len(&self) -> (l: usize)
+        ensures 
+            l as nat === self.view().len()
     {
-        self.inner.len()
+        self.inner.chars().count()
     }
-    
+    /// Warning: O(n) not O(1) due to unicode decoding needed
     #[verifier(external_body)]
-    pub fn get_char(&self, i: usize) -> (c: u8)
-        requires i < self.bytes().len()
+    pub fn get_char(&self, i: usize) -> (c: char)
+        requires i < self.view().len()
         ensures
-            self.is_ascii() ==> (
-                self.bytes().index(i as int) == c &&
-                c < 128
-            )
+            self.view().index(i as int) === c
     {
-        self.inner.as_bytes()[i]
+        self.inner.chars().nth(i).unwrap()
     }
 
     #[verifier(external_body)]
-    pub fn substring(&self, from: usize, to: usize) -> (ret: StrSlice<'a>)
+    pub fn substring_ascii(&self, from: usize, to: usize) -> (ret: StrSlice<'a>)
         requires
-            from < self.bytes().len(),
-            to <= self.bytes().len(),
+            self.is_ascii(),
+            from < self.view().len(),
+            to <= self.view().len(),
         ensures
-            ret.bytes() === self.bytes().subrange(from as int, to as int)
+            ret.view() === self.view().subrange(from as int, to as int)
     {
         StrSlice {
             inner: &self.inner[from..to],
         }
     }
 
+    #[verifier(external_body)]
+    pub fn substring_char(&self, from: usize, to: usize) -> (ret: StrSlice<'a>)
+        requires
+            from < self.view().len(),
+            to <= self.view().len()
+        ensures
+            ret.view() === self.view().subrange(from as int, to as int)
+    {
+        let mut char_pos = 0;
+        let mut byte_start = None;
+        let mut byte_end = None;
+        let mut byte_pos = 0;
+        let mut it = self.inner.chars();
+        loop {
+            if char_pos == from { byte_start = Some(byte_pos); }
+            if char_pos == to {
+                byte_end = Some(byte_pos);
+                break;
+            }
+
+            if let Some(c) = it.next() {
+                char_pos += 1;
+                byte_pos += c.len_utf8();
+            }
+            else { break; }
+        }
+        let byte_start = byte_start.unwrap();
+        let byte_end = byte_end.unwrap();
+        StrSlice {
+            inner: &self.inner[byte_start..byte_end]
+        }
+    }
+
     #[rustc_diagnostic_item = "pervasive::string::StrSlice::to_string"]
     #[verifier(external_body)]
```

Comment:
I didn't notice this before, but I don't think this needs to be `external_body` as it calls a function with a requires/ensures.